### ISSUE TITLE
Unexpected EOF while trying to create DNS record

### ISSUE THREAD

**Issue by SvenSchipper** [terraform-providers/terraform-provider-dns#24]
### Terraform Version

Terraform v0.11.1

### Affected Resource(s)

- dns_a_record_set

### Terraform Configuration Files

Note: names and addresses have been changed in the code snippet below 
The sku of the machine I'm trying to connect is 2012-R2-Datacenter-smalldisk

```hcl
provider "dns" {
  update {
    server = "10.10.10.10"
  }
}

resource "dns_a_record_set" "test" {
  zone = "example.com."
  name = "www."
  addresses = [
    "10.10.10.11",
  ]
}
```

### Debug Output

```
2018/01/09 10:48:15 [DEBUG] apply: dns_a_record_set.test: executing Apply
2018-01-09T10:48:15.285Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: panic: runtime error: invalid memory address or nil pointer dereference
2018-01-09T10:48:15.285Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x9dd9c3]
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: goroutine 55 [running]:
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/vendor/github.com/miekg/dns.(*Msg).Insert(0xc4202d7850, 0xc4202d7718, 0x1, 0x1)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/vendor/github.com/miekg/dns/update.go:71 +0xa3
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/dns.resourceDnsARecordSetUpdate(0xc420231960, 0xa9e020, 0xc420071360, 0x2, 0x2)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/dns/resource_dns_a_record_set.go:144 +0x89e
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/dns.resourceDnsARecordSetCreate(0xc420231960, 0xa9e020, 0xc420071360, 0x24, 0x1051b40)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/dns/resource_dns_a_record_set.go:58 +0x2f7
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc4202a33e0, 0xc4201110e0, 0xc42034a1a0, 0xa9e020, 0xc420071360, 0x1, 0xc42034c030, 0xc420280db0)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:193 +0x3b6
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc420231500, 0xc420111090, 0xc4201110e0, 0xc42034a1a0, 0x7f93826c06c8, 0x0, 0x18)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:259 +0xa4
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Apply(0xc4202e9960, 0xc4202e9d80, 0xc420320b10, 0x0, 0x0)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:488 +0x57
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: reflect.Value.call(0xc420130f00, 0xc420118be0, 0x13, 0xbc8b80, 0x4, 0xc420041f20, 0x3, 0x3, 0xc4200266e8, 0xc4200829c8, ...)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/usr/local/go/src/reflect/value.go:434 +0x906
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: reflect.Value.Call(0xc420130f00, 0xc420118be0, 0x13, 0xc420026720, 0x3, 0x3, 0xc400000005, 0x0, 0x0)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/usr/local/go/src/reflect/value.go:302 +0xa4
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: net/rpc.(*service).call(0xc42031c540, 0xc420110ff0, 0xc42011bb78, 0xc420129e80, 0xc4203232c0, 0xaa02a0, 0xc4202e9d80, 0x16, 0xaa02e0, 0xc420320b10, ...)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/usr/local/go/src/net/rpc/server.go:381 +0x142
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: created by net/rpc.(*Server).ServeCodec
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/usr/local/go/src/net/rpc/server.go:475 +0x36b
2018-01-09T10:48:15.286Z [DEBUG] plugin: plugin process exited: path=/home/qurrent/azureinfrastructure/azureinfrastructure/global/terraform-modules/test/.terraform/plugins/linux_amd64/terraform-provider-dns_v1.0.0_x4
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalWriteState
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalIf
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalWriteState
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalWriteDiff
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalApplyPost
2018/01/09 10:48:15 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* dns_a_record_set.test: unexpected EOF
2018/01/09 10:48:15 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* dns_a_record_set.test: unexpected EOF
2018/01/09 10:48:15 [TRACE] [walkApply] Exiting eval tree: dns_a_record_set.test

```
full output at:
https://gist.github.com/SvenSchipper/1b247f0cf1d1bf7a961453a24c2a344d

### Panic Output

full output at:
https://gist.github.com/SvenSchipper/0ffaab89f4386d8bc6e27a442a3e8fe8

### Expected Behavior

An A type DNS record set will be created.

### Actual Behavior
```
err: 1 error(s) occurred:

* dns_a_record_set.test: unexpected EOF
```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids
This is run in the Azure environment

**Comment by SFAntti** [comment]
I'm seeing the same issue while also having the key_name, key_algorithm and key_secret specified in the provider segment.

**Comment by SFAntti** [comment]
I found two issues in my configuration that were affecting my situation:
1. Try removing the dot from the end of the name field in dns_a_record_set. It seems like a dot is added automatically between the name and the zone.
2. I had specified a key_name in my provider config and it included a set of characters in uppercase. While looking at the embedded DNS library (github.com/miekg/dns) I noticed that the key_name was expected to be all lowercase. Switching to lowercase solved the issue.

After fixing these I was able to successfully register dns entries with the provider.  However, terraform-provider-dns should not crash, even with malformed input.
